File: app/src/main/java/com/quran/labs/androidquran/util/QuranAppUtils.java
package com.quran.labs.androidquran.util;

import android.text.TextUtils;

import com.quran.labs.androidquran.data.Constants;
import com.quran.labs.androidquran.data.QuranInfo;
import com.quran.labs.androidquran.data.SuraAyah;

import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.concurrent.Callable;

import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;

public class QuranAppUtils {

  public static Single<String> getQuranAppUrlObservable(final String key,
      final SuraAyah start, final SuraAyah end) {
    return Single.fromCallable(new Callable<String>() {
      @Override
      public String call() throws Exception {
        int sura = start.sura;
        int startAyah = start.ayah;
        // quranapp only supports sharing within a sura
        int endAyah = end.sura == start.sura ? end.ayah : QuranInfo.getNumAyahs(start.sura);
        return QuranAppUtils.getQuranAppUrl(key, sura, startAyah, endAyah);
      }
    }).subscribeOn(Schedulers.io());
  }

  private static String getQuranAppUrl(String key, int sura, Integer startAyah, Integer endAyah) {
    String url = null;
    String fallbackUrl = null;
    try {
      Map<String, String> params = new HashMap<String, String>();
      params.put("surah", sura + "");
      fallbackUrl = Constants.QURAN_APP_BASE + sura;
      if (startAyah != null) {
        params.put("start_ayah", startAyah.toString());
        fallbackUrl += "/" + startAyah;
        if (endAyah != null) {
          params.put("end_ayah", endAyah.toString());
          fallbackUrl += "-" + endAyah;
        } else {
          params.put("end_ayah", startAyah.toString());
        }
      }
      params.put("key", key);
      String result = getQuranAppUrl(params);
      if (!TextUtils.isEmpty(result)) {
        JSONObject json = new JSONObject(result);
        url = json.getString("url");
      }
    } catch (Exception e) {
      Timber.d(e, "error getting QuranApp url");
    }

    Timber.d("got back %s and fallback %s", url, fallbackUrl);
    return TextUtils.isEmpty(url) ? fallbackUrl : url;
  }

  private static String getQuranAppUrl(Map<String, String> params)
      throws IOException {
    URL url = null;
    try {
      url = new URL(Constants.QURAN_APP_ENDPOINT);
    } catch (MalformedURLException me) {
      // ignore
    }

    StringBuilder builder = new StringBuilder();
    Iterator<Map.Entry<String, String>> iterator =
        params.entrySet().iterator();
    while (iterator.hasNext()) {
      Map.Entry<String, String> item = iterator.next();
      builder.append(item.getKey()).append("=").append(item.getValue());
      if (iterator.hasNext()) {
        builder.append('&');
      }
    }

    String result = "";
    String body = builder.toString();
    byte[] bytes = body.getBytes();
    HttpURLConnection conn = null;
    try {
      // TODO: use OkHttp
      conn = (HttpURLConnection) url.openConnection();
      conn.setReadTimeout(10000);
      conn.setConnectTimeout(15000);
      conn.setDoOutput(true);
      conn.setDoInput(true);
      conn.setUseCaches(false);
      conn.setFixedLengthStreamingMode(bytes.length);
      conn.setRequestMethod("POST");
      conn.setRequestProperty("Content-Type",
          "application/x-www-form-urlencoded;charset=UTF-8");

      // post the request
      OutputStream out = conn.getOutputStream();
      out.write(bytes);
      out.close();

      // handle the response
      BufferedReader reader =
          new BufferedReader(new InputStreamReader(
              conn.getInputStream(), "UTF-8"));

      String line;
      while ((line = reader.readLine()) != null) {
        result += line;
      }

      try {
        reader.close();
      } catch (Exception e) {
        // ignore
      }
    } finally {
      if (conn != null) {
        conn.disconnect();
      }
    }

    return result;
  }
}

File: app/src/main/java/com/quran/labs/androidquran/util/QuranAppUtils.kt
package com.quran.labs.androidquran.util

import android.text.TextUtils
import com.quran.data.core.QuranInfo
import com.quran.data.model.SuraAyah
import com.quran.labs.androidquran.data.Constants
import dagger.Reusable
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.schedulers.Schedulers
import org.json.JSONObject
import timber.log.Timber
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL
import javax.inject.Inject

@Reusable
class QuranAppUtils @Inject internal constructor(private val quranInfo: QuranInfo) {
  fun getQuranAppUrlObservable(
    key: String,
    start: SuraAyah,
    end: SuraAyah
  ): Single<String> {
    return Single.fromCallable {
      val sura = start.sura
      val startAyah = start.ayah
      // quranapp only supports sharing within a sura
      val endAyah = if (end.sura == start.sura) end.ayah else quranInfo.getNumberOfAyahs(start.sura)
      getQuranAppUrl(key, sura, startAyah, endAyah)
    }.subscribeOn(Schedulers.io())
  }

  private fun getQuranAppUrl(key: String, sura: Int, startAyah: Int?, endAyah: Int?): String {
    val fallbackUrl = Constants.QURAN_APP_BASE + sura
    return try {
      val params = buildMap {
        put("surah", sura.toString())
        if (startAyah != null) {
          put("start_ayah", startAyah.toString())
          if (endAyah != null) {
            put("end_ayah", endAyah.toString())
          } else {
            put("end_ayah", startAyah.toString())
          }
        }
        put("key", key)
      }
      val result = getQuranAppUrl(params)
      val url = if (!TextUtils.isEmpty(result)) {
        val json = JSONObject(result)
        json.getString("url")
      } else {
        ""
      }

      return url.ifEmpty { fallbackUrl }
    } catch (e: Exception) {
      Timber.d(e, "error getting QuranApp url")
      fallbackUrl
    }
  }

  companion object {
    @Throws(IOException::class)
    private fun getQuranAppUrl(params: Map<String, String>): String {
      val url = try {
        URL(Constants.QURAN_APP_ENDPOINT)
      } catch (me: MalformedURLException) {
        // ignore
        null
      }

      val builder = StringBuilder()
      val iterator = params.entries.iterator()
      while (iterator.hasNext()) {
        val item = iterator.next()
        builder.append(item.key).append("=").append(item.value)
        if (iterator.hasNext()) {
          builder.append('&')
        }
      }

      var result = ""
      val body = builder.toString()
      val bytes = body.toByteArray(charset("UTF-8"))
      var conn: HttpURLConnection? = null
      try {
        // TODO: use OkHttp
        conn = (url!!.openConnection() as HttpURLConnection).apply {
          readTimeout = 10000
          connectTimeout = 15000
          doOutput = true
          doInput = true
          useCaches = false
          setFixedLengthStreamingMode(bytes.size)
          requestMethod = "POST"
          setRequestProperty(
            "Content-Type",
            "application/x-www-form-urlencoded;charset=UTF-8"
          )
        }

        // post the request
        val out = conn.outputStream
        out.write(bytes)
        out.close()

        // handle the response
        val reader =
          BufferedReader(
            InputStreamReader(
              conn.inputStream, "UTF-8"
            )
          )

        var line: String
        while ((reader.readLine().also { line = it }) != null) {
          result += line
        }

        try {
          reader.close()
        } catch (e: Exception) {
          // ignore
        }
      } finally {
        conn?.disconnect()
      }

      return result
    }
  }
}
